File: src/ios/ImageCompressor.swift
import UIKit

/// Downscales and JPEG-compresses captured selfie frames to a byte budget before they leave
/// the native layer.
///
/// Liveness frames are raw camera frames; handing one to the WebView as base64 and then on to
/// the back office is wasteful and slow. We crop to the face, cap the long edge, then step JPEG
/// quality down until the payload fits the budget.
///
/// ImageCompressor.java mirrors this behaviour — keep the two in sync.
final class ImageCompressor {

    /// Quality step used when walking down towards the byte budget.
    private static let qualityStep: CGFloat = 0.08

    struct Options {
        /// Longest edge of the output image, in pixels.
        var maxDimension: Int = 720
        /// Hard budget for the encoded JPEG, in bytes.
        var maxBytes: Int = 200 * 1024
        var initialQuality: CGFloat = 0.85
        var minQuality: CGFloat = 0.45
        /// Crop to the face box, expanded by `faceCropPadding` of the box on each side.
        var cropToFace: Bool = true
        var faceCropPadding: CGFloat = 0.55
        /// Flip the output horizontally.
        ///
        /// Detection runs on a mirrored image (see LivenessDetector.swift), so face boxes align
        /// with a mirrored frame — but the portrait we hand to the back office should be in true
        /// orientation to match the document portrait. We therefore crop in mirrored space and
        /// flip once at the end. Android needs no flip: its analysis frames are never mirrored.
        var mirrorHorizontally: Bool = true
    }

    struct Result {
        let data: Data
        let width: Int
        let height: Int
        let quality: CGFloat

        /// Base64 without line wrapping, matching the faceImageBase64 field from the NFC chip.
        var base64: String {
            return data.base64EncodedString()
        }
    }

    /// - Parameters:
    ///   - image: the captured frame, already upright in the same space ML Kit reported boxes in
    ///   - faceBox: face bounds in `image` pixel coordinates, or nil for no crop
    static func compress(_ image: UIImage, faceBox: CGRect?, options: Options) -> Result? {
        guard let source = image.cgImage else {
            NSLog("[ImageCompressor] Frame has no backing CGImage — skipping")
            return nil
        }

        var working = source
        if options.cropToFace, let faceBox = faceBox,
           let cropped = crop(source, to: faceBox, padding: options.faceCropPadding) {
            working = cropped
        }

        let targetSize = scaledSize(width: working.width,
                                    height: working.height,
                                    maxDimension: options.maxDimension)

        guard let rendered = render(working,
                                    to: targetSize,
                                    mirrored: options.mirrorHorizontally) else {
            return nil
        }

        var quality = min(max(options.initialQuality, options.minQuality), 1.0)
        guard var data = rendered.jpegData(compressionQuality: quality) else {
            return nil
        }

        while data.count > options.maxBytes && quality > options.minQuality {
            quality = max(options.minQuality, quality - qualityStep)
            guard let next = rendered.jpegData(compressionQuality: quality) else { break }
            data = next
        }

        // Size only — never log image bytes or base64: this is biometric PII.
        NSLog("[ImageCompressor] Compressed selfie: %dx%d q=%.2f bytes=%d%@",
              Int(targetSize.width), Int(targetSize.height), Double(quality), data.count,
              data.count > options.maxBytes ? " (over budget at min quality)" : "")

        return Result(data: data,
                      width: Int(targetSize.width),
                      height: Int(targetSize.height),
                      quality: quality)
    }

    // MARK: - Steps

    /// Expands the ML Kit face box outwards so the crop keeps hair, chin and some background —
    /// face matchers do better with that context than with a box cropped tight to the features.
    private static func crop(_ source: CGImage, to faceBox: CGRect, padding: CGFloat) -> CGImage? {
        let padX = faceBox.width * padding
        let padY = faceBox.height * padding

        let expanded = faceBox.insetBy(dx: -padX, dy: -padY)
        let bounds = CGRect(x: 0, y: 0, width: source.width, height: source.height)
        let clamped = expanded.intersection(bounds).integral

        guard !clamped.isNull, clamped.width >= 1, clamped.height >= 1 else {
            NSLog("[ImageCompressor] Face box outside frame bounds — skipping crop")
            return nil
        }
        return source.cropping(to: clamped)
    }

    private static func scaledSize(width: Int, height: Int, maxDimension: Int) -> CGSize {
        let longEdge = max(width, height)
        guard maxDimension > 0, longEdge > maxDimension else {
            return CGSize(width: width, height: height)
        }
        let scale = CGFloat(maxDimension) / CGFloat(longEdge)
        return CGSize(width: max(1, (CGFloat(width) * scale).rounded()),
                      height: max(1, (CGFloat(height) * scale).rounded()))
    }

    private static func render(_ source: CGImage, to size: CGSize, mirrored: Bool) -> UIImage? {
        let format = UIGraphicsImageRendererFormat()
        // Pixel-for-pixel output: the caller asked for a specific pixel budget, not points.
        format.scale = 1
        format.opaque = true

        let renderer = UIGraphicsImageRenderer(size: size, format: format)
        return renderer.image { context in
            if mirrored {
                context.cgContext.translateBy(x: size.width, y: 0)
                context.cgContext.scaleBy(x: -1, y: 1)
            }
            UIImage(cgImage: source).draw(in: CGRect(origin: .zero, size: size))
        }
    }
}

File: src/android/ImageCompressor.java
package com.nfcdocumentreader;

import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.Base64;
import android.util.Log;

import java.io.ByteArrayOutputStream;

/**
 * Downscales and JPEG-compresses captured selfie frames to a byte budget before they leave
 * the native layer.
 *
 * Liveness frames are raw camera frames (1280x720 ARGB is ~3.5 MB); handing that to the
 * WebView as base64 and then to the back office is wasteful and slow. We crop to the face,
 * cap the long edge, then step JPEG quality down until the payload fits the budget.
 *
 * ImageCompressor.swift mirrors this behaviour — keep the two in sync.
 */
public class ImageCompressor {

    private static final String TAG = "ImageCompressor";

    /** Quality step used when walking down towards the byte budget. */
    private static final int QUALITY_STEP = 8;

    public static class Options {
        /** Longest edge of the output image, in pixels. */
        public int maxDimension = 720;
        /** Hard budget for the encoded JPEG, in bytes. */
        public int maxBytes = 200 * 1024;
        public int initialQuality = 85;
        public int minQuality = 45;
        /** Crop to the face box, expanded by this fraction of the box on each side. */
        public boolean cropToFace = true;
        public float faceCropPadding = 0.55f;
    }

    public static class Result {
        public byte[] jpeg;
        public int width;
        public int height;
        public int quality;

        /** Base64 without line wrapping, matching the faceImageBase64 field from the NFC chip. */
        public String toBase64() {
            return Base64.encodeToString(jpeg, Base64.NO_WRAP);
        }
    }

    /**
     * @param source  the captured frame, already rotated upright
     * @param faceBox face bounds in {@code source} pixel coordinates, or null for no crop
     */
    public static Result compress(Bitmap source, Rect faceBox, Options options) {
        Bitmap working = source;
        Bitmap cropped = null;
        Bitmap scaled = null;

        try {
            if (options.cropToFace && faceBox != null) {
                cropped = cropToFace(source, faceBox, options.faceCropPadding);
                if (cropped != null) {
                    working = cropped;
                }
            }

            scaled = scaleToMaxDimension(working, options.maxDimension);
            if (scaled != null) {
                working = scaled;
            }

            Result result = new Result();
            result.width = working.getWidth();
            result.height = working.getHeight();

            int quality = clamp(options.initialQuality, options.minQuality, 100);
            byte[] encoded = encode(working, quality);

            while (encoded.length > options.maxBytes && quality > options.minQuality) {
                quality = Math.max(options.minQuality, quality - QUALITY_STEP);
                encoded = encode(working, quality);
            }

            result.jpeg = encoded;
            result.quality = quality;

            // Size only — never log image bytes or base64: this is biometric PII.
            Log.d(TAG, "Compressed selfie: " + result.width + "x" + result.height
                    + " q=" + quality + " bytes=" + encoded.length
                    + (encoded.length > options.maxBytes ? " (over budget at min quality)" : ""));

            return result;
        } finally {
            // Recycle only the intermediates we created — `source` belongs to the caller.
            // Both are safe to drop here: the JPEG bytes are already encoded.
            if (scaled != null && scaled != source) {
                scaled.recycle();
            }
            if (cropped != null && cropped != source && cropped != scaled) {
                cropped.recycle();
            }
        }
    }

    // ==================== Steps ====================

    /**
     * Expands the ML Kit face box outwards so the crop keeps hair, chin and some background —
     * face matchers do better with that context than with a box cropped tight to the features.
     */
    private static Bitmap cropToFace(Bitmap source, Rect faceBox, float padding) {
        int padX = (int) (faceBox.width() * padding);
        int padY = (int) (faceBox.height() * padding);

        int left = Math.max(0, faceBox.left - padX);
        int top = Math.max(0, faceBox.top - padY);
        int right = Math.min(source.getWidth(), faceBox.right + padX);
        int bottom = Math.min(source.getHeight(), faceBox.bottom + padY);

        int width = right - left;
        int height = bottom - top;
        if (width <= 0 || height <= 0) {
            Log.w(TAG, "Face box outside frame bounds — skipping crop");
            return null;
        }

        return Bitmap.createBitmap(source, left, top, width, height);
    }

    private static Bitmap scaleToMaxDimension(Bitmap source, int maxDimension) {
        int longEdge = Math.max(source.getWidth(), source.getHeight());
        if (maxDimension <= 0 || longEdge <= maxDimension) {
            return null;
        }

        float scale = (float) maxDimension / longEdge;
        int width = Math.max(1, Math.round(source.getWidth() * scale));
        int height = Math.max(1, Math.round(source.getHeight() * scale));
        return Bitmap.createScaledBitmap(source, width, height, true);
    }

    private static byte[] encode(Bitmap bitmap, int quality) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, quality, out);
        return out.toByteArray();
    }

    private static int clamp(int value, int min, int max) {
        return Math.max(min, Math.min(max, value));
    }
}
